flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartition.java

```diff
@@ -289,8 +374,20 @@ public void finish() throws IOException {
         super.finish();
     }
 
+    private void releaseWriteBuffers() {
+        synchronized (lock) {
+            if (bufferPool != null) {
+                for (MemorySegment segment : writeBuffers) {
+                    bufferPool.recycle(segment);
+                }
+                writeBuffers.clear();
+            }
+        }
+    }
+
     @Override
     public void close() {
+        releaseWriteBuffers();
         releaseCurrentSortBuffer();
```

Comment:
Although not fully related with this PR, currently the sort buffer seems to have different treatment with the write buffers on lock. We might add a comment on there would always be a close() call from the main thread or we might change currentSortBuffer to be volatile.